eth/local_tx_handler.go

```diff
@@ -0,0 +1,497 @@
+
+// mergeNonceOrderedTxs combines curTxs and newTxs together.  each list is nonce-contiguous, nonce-unique ordered ascending by nonce.  If the nonce of a transaction in newTxs matches a transaction in curTxs, only the transaction from newTxs is included in the result.
+// The resulting list is nonce-contiguous, nonce-unique and ordered ascending by nonce.
+func mergeNonceOrderedTxs(curTxs []*types.Transaction, newTxs []*types.Transaction) []*types.Transaction {
+	// insert new txs into the sender's queue, keeping the resulting array nonce-ordered and
+	// replacing pre-existing txs if there is a tx with same nonce
+	curTxsIdx, newTxsIdx := 0, 0
+	var res []*types.Transaction
+
+	for curTxsIdx < len(curTxs) || newTxsIdx < len(newTxs) {
+		if curTxsIdx < len(curTxs) && newTxsIdx < len(newTxs) {
+			if curTxs[curTxsIdx].Nonce() < newTxs[newTxsIdx].Nonce() {
+				res = append(res, curTxs[curTxsIdx])
+				curTxsIdx++
+			} else if curTxs[curTxsIdx].Nonce() > newTxs[newTxsIdx].Nonce() {
+				res = append(res, newTxs[newTxsIdx])
+				newTxsIdx++
+			} else {
+				res = append(res, newTxs[newTxsIdx])
+				curTxsIdx++
+				newTxsIdx++
+			}
+		} else if curTxsIdx < len(curTxs) {
+			res = append(res, curTxs[curTxsIdx])
+			curTxsIdx++
+		} else if newTxsIdx < len(newTxs) {
+			res = append(res, newTxs[newTxsIdx])
+			newTxsIdx++
+		}
+	}
+	return res
+}
+
+// updateLowestUnsentNonce sets the lowest unsent nonce for the account in each peer's tracker if
+// the new lowest unsent nonce is lte the current one for the peer/sender.
+func (l *localTxHandler) updateLowestUnsentNonce(sender common.Address, nonce uint64) {
+	for _, peer := range l.peers.allEthPeers() {
+		lastUnsentNonce := l.peersStatus.getLowestUnsentNonce(peerID(peer.ID()), sender)
+		if nonce <= lastUnsentNonce {
+			l.peersStatus.setLowestUnsentNonce(peerID(peer.ID()), sender, nonce)
+		}
+	}
+}
+
+// updateLowestUnannouncedNonce sets the lowest unannounced nonce for the account in each peer's tracker if
+// the new lowest unsent nonce is lte the current one for the peer/sender.
+func (l *localTxHandler) updateLowestUnannouncedNonce(sender common.Address, nonce uint64) {
+	for _, peer := range l.peers.allEthPeers() {
+		lastUnsentNonce := l.peersStatus.getLowestUnannouncedNonce(peerID(peer.ID()), sender)
+		if nonce <= lastUnsentNonce {
+			l.peersStatus.setLowestUnannouncedNonce(peerID(peer.ID()), sender, nonce)
+		}
+	}
+}
+
+// addLocalsFromSender inserts a list of nonce-ordered transactions into the tracking
+// queue for the associated account.
+func (l *localTxHandler) addLocalsFromSender(sender common.Address, newTxs []*types.Transaction) {
+	var curTxs []*types.Transaction
+
+	curTxs, ok := l.txQueues[sender]
+	if !ok {
+		l.txQueues[sender] = newTxs
+		l.updateLowestUnsentNonce(sender, newTxs[0].Nonce())
+		l.updateLowestUnannouncedNonce(sender, newTxs[0].Nonce())
+		return
+	}
+
+	newTxsProceedCurrent := newTxs[0].Nonce() > l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()
```

Comment:
```code
	newTxsProceedCurrent := newTxs[0].Nonce() > l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()
```
Wow, what a mouthful: `l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()`. 

```suggestion
	newTxsProceedCurrent := newTxs[0].Nonce() > curTxs[len(curTxs) -1].Nonce()
```